tensorboard/components/tf_dashboard_common/data-loader-behavior.ts

```diff
@@ -149,43 +164,62 @@ namespace tf_dashboard_common {
       this.cancelAsync(this._loadDataAsync);
-      this._loadDataAsync = this.async(() => {
-        // Read-only property have a special setter.
-        this._setDataLoading(true);
-
-        // Before updating, cancel any network-pending updates, to
-        // prevent race conditions where older data stomps newer data.
-        this._canceller.cancelAll();
-        const promises = this.dataToLoad
-          .filter((datum) => {
-            const name = this.getDataLoadName(datum);
-            return !this._inflightOrLoadedData.has(name);
-          })
-          .map((datum) => {
-            const cacheKey = this.getDataLoadName(datum);
-            this._inflightOrLoadedData.add(cacheKey);
-            return this.requestData(datum).then((value) => {
-              // dataToLoad can change while the request in-flight and it may
-              // not be no longer required (loadDataCallback can be an expensive
-              // proess so we would like to omit it if possible).
-              const isDataRequiredNow = this.dataToLoad.find(
-                (datum) => this.getDataLoadName(datum) === cacheKey
+      this._loadDataAsync = this.async(
+        this._canceller.cancellable((result) => {
+          if (result.cancelled) {
+            return;
+          }
+          // Read-only property have a special setter.
+          this._setDataLoading(true);
+
+          // Promises return cacheKeys of the data that were fetched.
+          const promises = this.dataToLoad
+            .filter((datum) => {
+              const cacheKey = this.getDataLoadName(datum);
+              return !this._dataLoadState.has(cacheKey);
+            })
+            .map((datum) => {
+              const cacheKey = this.getDataLoadName(datum);
+              this._dataLoadState.set(cacheKey, LoadState.LOADING);
+              return this.requestData(datum).then(
+                this._canceller.cancellable((result) => {
+                  // It was resetted. Do not notify of the response.
+                  if (!result.cancelled) {
+                    this._dataLoadState.set(cacheKey, LoadState.LOADED);
+                    this.loadDataCallback(this, datum, result.value);
+                  }
+                  return cacheKey;
+                })
               );
-              if (isDataRequiredNow) {
-                this.loadDataCallback(this, datum, value);
-              }
             });
-          });
-
-        return Promise.all(promises).then(
-          this._canceller.cancellable((result) => {
-            // Read-only property have a special setter.
-            this._setDataLoading(false);
-            if (result.cancelled) return;
-            this.onLoadFinish();
-          })
-        );
-      });
+
+          return Promise.all(promises).then(
+            this._canceller.cancellable((result) => {
+              // It was resetted. Do not notify of the data load.
+              if (!result.cancelled) {
```

Comment:
veeery small nit: I prefer early returns, to prevent deeper indentation

Reply:
Thanks for the review. Since our CI takes super long and we need this to unblock sync, I will merge as is and we can address this in future time if it truly bothers you.